### DISCUSSION TITLE
micropython/lora , modem.poll_recv() in a callback problem

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#13341] [Libraries & Drivers]
I have used `start_recv` and `poll_recv` in some low-level API code to successfully get the messages send from a remote.

When I tried using an `irq_my_rx()` callback I determined that, at least `poll_recv` would get me the message.  However, doing a `decode` on the data throws a `unicode error`.

I also tried:
```
py
sender_id, my_data = my_receiver.recv(wait=True)
```
in `my_rx_ irq()`  gives `Invalid length in payload header`

Writing a Python script to decode the message I discovered that there is a start byte in my message.  When I remove it the message decodes.
```py
#my_data = bytearray(b'\xfa\x01}\x18Hello, ticks_ms=36083677\x12')
my_data = bytearray(b'\x01}\x18Hello, ticks_ms=36083677\x12')
print(type(my_data))
print(my_data)
my_data = my_data.decode('utf-8')
print(type(my_data))
print(my_data)
```
What is the correct procedure to recover the received data in a callback?

**Comment by sophiedegran** [comment]
Perharps you have to 'reset' the lora buffer before reading ?

**Reply by davefes** [reply]
Good suggestion but I don't see any methods to do that.  I will look harder in the source code ... as Googling for ways to remove a encoded start byte (xfa) out of a byetarray has not been fruitful.

**Reply by davefes** [reply]
According to this snippet from `sender.py`
```py
        # Prepare the simple payload with header and checksum
        # See README for a summary of the simple data message format
        payload = bytearray(len(sensor_data) + 5)
        struct.pack_into("<HBB", payload, 0, self.device_id, self.counter, len(sensor_data))
        payload[4:-1] = sensor_data
        payload[-1] = sum(b for b in payload) & 0xFF
```
I am probably seeing a lot more than the original `sensor_data`.  Looks like post-processing is required after a `poll_recv()`
